### PR TITLE
"Data Picker" editor

### PR DESCRIPTION
**Pull request by leekelleher**
### Description

Adds a **Data Picker** editor.

_(Screen capture video demo coming soon.)_

### Types of changes

- [x] Documentation change
- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [x] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_

### Checklist

- [x] My code follows the coding style of this project.
- [x] My changes generate no new warnings.
- [x] My change requires a change to the documentation.
- [x] I have updated the corresponding documentation.
- [x] I have read the **[CONTRIBUTING](CONTRIBUTING)** and **[CODE_OF_CONDUCT](CODE_OF_CONDUCT.md)** documents.

### CODE REVIEW

File: src/Umbraco.Community.Contentment/DataEditors/DataPicker/DataPickerApiController.cs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+            IDataTypeService dataTypeService,
+            ConfigurationEditorUtility utility)
+        {
+            _dataTypeService = dataTypeService;
+            _utility = utility;
+        }
+
+        [HttpPost]
+#if NET472
+        public HttpResponseMessage GetItems(Guid dataTypeKey, [FromBody] string[] values)
```
Comment (nzdev) [inline src/Umbraco.Community.Contentment/DataEditors/DataPicker/DataPickerApiController.cs:50]: Could controllers be async? This way it's easier to use async data sources.

**Comment by nzdev** [comment]
image
For no url content, the udi overflows on the cards view

**Comment by nzdev** [comment]
I'd be very keen to have support (and I'm willing to contribute) for multiple inputs to add into the query so I can generate queries with additional parameters like a date range, a drop down list, checkboxes (possibly with facet counts), picking a node from the content / media tree, and other dynamically loading parameters. Prime scenario being calling rest apis and running examine queries.

**Comment by nzdev** [comment]
Sorry if this is too early in the process for feedback, setting the overlay size in the data picker source does not change the overlay size.

File: src/Umbraco.Community.Contentment/DataEditors/DataPicker/DataPickerApiController.cs (hunk 1 above)
Comment (nzdev) [reply src/Umbraco.Community.Contentment/DataEditors/DataPicker/DataPickerApiController.cs:50]: PR https://github.com/leekelleher/umbraco-contentment/pull/298

**Comment by leekelleher** [comment]
@nzdev Thank you for the feedback so far - and for trying it out! 🙏

Once tomorrow's [Umbraco Community Day](https://day.umbraco.community/) session is done, I'll review and get back to you. 😀

**Comment by nzdev** [comment]
Thanks @leekelleher . Both of these suggestions I would see being done the same as the built in list views property configuration.

**Comment by nzdev** [comment]
`Task<IEnumerable<DataPickerItem>> SearchAsync(Dictionary<string, object> config, out int totalPages, int pageNumber = 1, int pageSize = 12, string query = "")`

This method on IDataPickerSource needs to change it's return type to include total pages as out parameters cannot be used with async.

**Comment by nzdev** [comment]
https://github.com/leekelleher/umbraco-contentment/pull/303

**Comment by leekelleher** [comment]
FYI, I've done a big refactor on how the data-sources for Data Picker work. I've decided to try to combine them with the existing Data List data-sources, as it felt that I was duplicating logic, e.g. the Umbraco Content data-source was largely similar to the Data List one.

**Comment by leekelleher** [comment]
I've decided to merge this in to the `dev/v4.x` branch for the next release, Contentment 4.5.0.

I haven't written any docs yet, but will do before the release (even if they are basic/light on detail).

There are a bunch of configuration options that I wanted to add in for Data Picker, e.g. fallback icon; enable/disable filter query; allow duplicates; enable multiple; confirm removals; et al.  These can be added in future release, otherwise I'll end up never releasing the Data Picker!

My last niggle is that the `data-picker.overlay.html` UI is tied exclusively to the "cards" and "list" UIs. I'd have liked to make this agnostic, but required much more dev time, so again, yet another thing to delay the initial release (where the majority of users may not notice this limitation).

There's only so long I could wait - given I opened this PR back in January! **Just ship it Lee!** :shipit:

**Comment by leekelleher** [comment]
@sannishad This very issue was raised last week, see #420 for all details. In short, the Examine data-source needs further development, which [is open for community contribution](https://github.com/leekelleher/umbraco-contentment/issues/421).

Please continue any further discussion on #420, and not this pull request for the original Data Picker editor, thank you. 🙏

**Pull request merged**